invokeai/app/services/shared/sqlite_migrator/sqlite_migrator_impl.py

```diff
@@ -0,0 +1,212 @@
+    """
+
+    backup_path: Optional[Path] = None
+
+    def __init__(
+        self,
+        db_path: Path | None,
+        conn: sqlite3.Connection,
+        lock: threading.RLock,
+        logger: Logger,
+        log_sql: bool = False,
+    ) -> None:
+        self._lock = lock
+        self._db_path = db_path
+        self._logger = logger
+        self._conn = conn
+        self._log_sql = log_sql
+        self._cursor = self._conn.cursor()
+        self._migration_set = MigrationSet()
+
+        # The presence of an temp database file indicates a catastrophic failure of a previous migration.
+        if self._db_path and self._get_temp_db_path(self._db_path).is_file():
+            self._logger.warning("Previous migration failed! Trying again...")
+            self._get_temp_db_path(self._db_path).unlink()
+
+    def register_migration(self, migration: Migration) -> None:
+        """Registers a migration."""
+        self._migration_set.register(migration)
+        self._logger.debug(f"Registered migration {migration.from_version} -> {migration.to_version}")
+
+    def run_migrations(self) -> bool:
+        """Migrates the database to the latest version."""
+        with self._lock:
+            # This throws if there is a problem.
+            self._migration_set.validate_migration_chain()
+            self._create_migrations_table(cursor=self._cursor)
+
+            if self._migration_set.count == 0:
+                self._logger.debug("No migrations registered")
+                return False
+
+            if self._get_current_version(self._cursor) == self._migration_set.latest_version:
+                self._logger.debug("Database is up to date, no migrations to run")
+                return False
+
+            self._logger.info("Database update needed")
+
+            if self._db_path:
+                # We are using a file database. Create a copy of the database to run the migrations on.
+                temp_db_path = self._create_temp_db(self._db_path)
+                self._logger.info(f"Copied database to {temp_db_path} for migration")
+                temp_db_conn = sqlite3.connect(temp_db_path)
+                # We have to re-set this because we just created a new connection.
+                if self._log_sql:
+                    temp_db_conn.set_trace_callback(self._logger.debug)
+                temp_db_cursor = temp_db_conn.cursor()
+                self._run_migrations(temp_db_cursor)
+                # Close the connections, copy the original database as a backup, and move the temp database to the
+                # original database's path.
+                temp_db_conn.close()
+                self._conn.close()
+                backup_db_path = self._finalize_migration(
+                    temp_db_path=temp_db_path,
+                    original_db_path=self._db_path,
+                )
+                self._logger.info(f"Migration successful. Original DB backed up to {backup_db_path}")
```

Comment:
SQLite is ACID-compliant and has [very robust failure handling](https://www.sqlite.org/transactional.html).

I'm questioning if our file shuffling is really adding any significant protection, because it definitely has some inconveniences:
- Backup DB files will accumulate as we run more migrations (unless I've missed the cleanup code somewhere?).
- If there is a failure during the final file renaming step, manual action could be required to recover (albeit unlikely)

Reply:
lol what a docs page. Handling files has been a pain, requiring me to boot up windows *hurk* to fix tests. I trust SQLite far more than myself, manual shuffling files. I'm happy to rely on SQLite.

Reply:
I've removed all file stuff from the migrator. I have trust in SQLite.